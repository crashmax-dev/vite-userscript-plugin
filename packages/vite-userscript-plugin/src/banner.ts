import type { MetadataConfig } from './types.js'

export function banner(config: MetadataConfig): string {
  const metadata: string[] = []
  const configKeys = Object.keys(config)
  const maxKeyLength = Math.max(...configKeys.map((key) => key.length)) + 1

  const addSpaces = (str: string): string => {
    return ' '.repeat(maxKeyLength - str.length)
  }

  const addMetadata = (key: string, value: string | number | boolean): void => {
    const isBoolean = typeof value === 'boolean'
    if (isBoolean && !value) return
    value = !isBoolean ? addSpaces(key) + value.toString() : ''
    metadata.push(`// @${key}${value}`)
  }

  for (const [key, value] of Object.entries(config)) {
    if (Array.isArray(value)) {
      value.forEach((value) => addMetadata(key, value))
    } else {
      if (value === undefined) continue
      addMetadata(key, value)
    }
  }

  return [
    '// ==UserScript==',
    ...metadata,
    '// ==/UserScript=='
  ].join('\n')
}
