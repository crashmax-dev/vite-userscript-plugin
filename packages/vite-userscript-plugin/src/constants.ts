export const regexpScripts = new RegExp(/\.js/)
export const regexpStyles = new RegExp(/\.css|\.sass|\.scss/)
export const template = `console.warn("__TEMPLATE__")`

export const grants = [
  'unsafeWindow',
  'window.onurlchange',
  'window.focus',
  'window.close',
  'GM_setValue',
  'GM_getValue',
  'GM_deleteValue',
  'GM_listValues',
  'GM_setClipboard',
  'GM_addStyle',
  'GM_addElement',
  'GM_addValueChangeListener',
  'GM_removeValueChangeListener',
  'GM_registerMenuCommand',
  'GM_unregisterMenuCommand',
  'GM_download',
  'GM_getTab',
  'GM_getTabs',
  'GM_saveTab',
  'GM_openInTab',
  'GM_notification',
  'GM_getResourceURL',
  'GM_getResourceText',
  'GM_xmlhttpRequest',
  'GM_webRequest',
  'GM_log',
  'GM_info'
] as const
