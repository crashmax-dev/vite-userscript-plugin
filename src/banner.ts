import type { HeaderConfig } from './types.js'

export function banner(config: HeaderConfig): string {
  const header: string[] = []
  const configKeys = Object.keys(config)
  const maxKeyLength = Math.max(...configKeys.map((key) => key.length)) + 1

  const addSpaces = (str: string): string => {
    return ' '.repeat(maxKeyLength - str.length)
  }

  const addMetadata = (key: string, value: string | number | boolean): void => {
    const isBoolean = typeof value === 'boolean'
    if (isBoolean && !value) return
    value = !isBoolean ? addSpaces(key) + value.toString() : ''
    header.push(`// @${key}${value}`)
  }

  for (const [key, value] of Object.entries(config)) {
    if (Array.isArray(value)) {
      value.forEach((value) => addMetadata(key, value))
    } else {
      if (value === undefined) continue
      addMetadata(key, value)
    }
  }

  return [
    '// ==UserScript==',
    ...header,
    '// ==/UserScript=='
  ].join('\n')
}
