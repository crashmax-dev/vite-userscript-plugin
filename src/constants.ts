export const regexpScripts = new RegExp(/.(t|j)sx?/)
export const regexpStyles = new RegExp(/\.(s?css|sass)$/)

export const grants = [
  'unsafeWindow',
  'window.onurlchange',
  'window.focus',
  'window.close',
  'GM_setValue',
  'GM_getValue',
  'GM_deleteValue',
  'GM_listValues',
  'GM_setClipboard',
  'GM_addStyle',
  'GM_addElement',
  'GM_addValueChangeListener',
  'GM_removeValueChangeListener',
  'GM_registerMenuCommand',
  'GM_unregisterMenuCommand',
  'GM_download',
  'GM_getTab',
  'GM_getTabs',
  'GM_saveTab',
  'GM_openInTab',
  'GM_notification',
  'GM_getResourceURL',
  'GM_getResourceText',
  'GM_xmlhttpRequest',
  'GM_webRequest',
  'GM_log',
  'GM_info'
] as const
