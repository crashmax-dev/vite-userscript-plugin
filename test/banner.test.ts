import { expect, test } from 'vitest'
import { banner } from '../src/banner.js'
import { grants } from '../src/constants.js'
import type { Grants, MetadataConfig } from '../src/types.js'

const metadataConfig: MetadataConfig = {
  name: 'vitest',
  version: '1.0.0',
  author: 'John Doe',
  description: 'vitest',
  namespace: 'vitest',
  connect: 'vitest.dev',
  license: 'MIT',
  noframes: true,
  icon: 'https://vitest.dev/favicon.ico',
  icon64: 'https://vitest.dev/favicon.ico',
  exclude: ['https://vitest.dev/guide/*', 'https://vitest.dev/api/*'],
  include: 'https://vitest.dev',
  homepage: 'https://github.com/vitest-dev/vitest',
  downloadURL: 'https://vitest.dev',
  supportURL: 'https://vitest.dev',
  updateURL: 'https://vitest.dev',
  resource: 'https://vitest.dev',
  require: 'https://example.com/index.js',
  grant: grants as unknown as Exclude<Grants, 'GM_addStyle' | 'GM_info'>[],
  match: 'https://vitest.dev',
  'run-at': 'document-start'
}

test('banner snapshot', () => {
  const defaultBanner = banner(metadataConfig)
  expect(defaultBanner).toMatchSnapshot()
})
