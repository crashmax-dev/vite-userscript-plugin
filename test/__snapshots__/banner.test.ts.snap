// Vitest Snapshot v1

exports[`banner default snapshot 1`] = `
"// ==UserScript==
// @name        vitest
// @version     1.0.0
// @author      John Doe
// @description vitest
// @namespace   vitest
// @connect     vitest.dev
// @license     MIT
// @noframes
// @icon        https://vitest.dev/favicon.ico
// @icon64      https://vitest.dev/favicon.ico
// @exclude     https://vitest.dev/guide/*
// @exclude     https://vitest.dev/api/*
// @include     https://vitest.dev
// @homepage    https://github.com/vitest-dev/vitest
// @downloadURL https://vitest.dev
// @supportURL  https://vitest.dev
// @updateURL   https://vitest.dev
// @resource    https://vitest.dev
// @require     https://example.com/index.js
// @grant       unsafeWindow
// @grant       window.onurlchange
// @grant       window.focus
// @grant       window.close
// @grant       GM_setValue
// @grant       GM_getValue
// @grant       GM_deleteValue
// @grant       GM_listValues
// @grant       GM_setClipboard
// @grant       GM_addStyle
// @grant       GM_addElement
// @grant       GM_addValueChangeListener
// @grant       GM_removeValueChangeListener
// @grant       GM_registerMenuCommand
// @grant       GM_unregisterMenuCommand
// @grant       GM_download
// @grant       GM_getTab
// @grant       GM_getTabs
// @grant       GM_saveTab
// @grant       GM_openInTab
// @grant       GM_notification
// @grant       GM_getResourceURL
// @grant       GM_getResourceText
// @grant       GM_xmlhttpRequest
// @grant       GM_webRequest
// @grant       GM_log
// @grant       GM_info
// @match       https://vitest.dev
// @run-at      document-start
// ==/UserScript=="
`;

exports[`banner meta snapshot 1`] = `
"// ==UserScript==
// @name        vitest
// @version     1.0.0
// @match       https://example.com
// @homepage    https://crashmax-dev.github.io/jsx/
// @updateURL   https://crashmax-dev.github.io/jsx/vitest.meta.js
// @downloadURL https://crashmax-dev.github.io/jsx/vitest.user.js
// ==/UserScript=="
`;
